modules/rgbd/src/depth_registration.cpp

```diff
@@ -0,0 +1,381 @@
+    isEqualToNoDepthSentinelValue(const DepthDepth &value)
+    {
+        return value == noDepthSentinelValue<DepthDepth>();
+    }
+
+    template<>
+    inline bool
+    isEqualToNoDepthSentinelValue<float>(const float &value)
+    {
+        return cvIsNaN(value) != 0;
+    }
+
+    template<>
+    inline bool
+    isEqualToNoDepthSentinelValue<double>(const double &value)
+    {
+        return cvIsNaN(value) != 0;
+    }
+
+ ///////////////////////////////////////////////////////////////////////////////////
+
+
+    // When using the unsigned short representation, we'd like to round the values to the nearest
+    // integer value. The float/double representations don't need to be rounded
+    template<typename DepthDepth>
+    inline DepthDepth
+    floatToInputDepth(const float &value)
+    {
+        return (DepthDepth)value;
+    }
+
+    template<>
+    inline unsigned short
+    floatToInputDepth<unsigned short>(const float &value)
+    {
+        return (unsigned short)(value+0.5);
+    }
+
+ ///////////////////////////////////////////////////////////////////////////////////
+
+
+    /** Computes a registered depth image from an unregistered image.
+     *
+     * @param unregisteredDepth the input depth data
+     * @param unregisteredCameraMatrix the camera matrix of the depth camera
+     * @param registeredCameraMatrix the camera matrix of the external camera
+     * @param registeredDistCoeffs the distortion coefficients of the external camera
+     * @param rbtRgb2Depth the rigid body transform between the cameras.
+     * @param outputImagePlaneSize the image plane dimensions of the external camera (width, height)
+     * @param depthDilation whether or not the depth is dilated to avoid holes and occlusion errors
+     * @param inputDepthToMetersScale the scale needed to transform the input depth units to meters
+     * @param registeredDepth the result of transforming the depth into the external camera
+     */
+    template<typename DepthDepth>
+    void
+    performRegistration(const Mat_<DepthDepth> &unregisteredDepth,
+                        const Matx33f &unregisteredCameraMatrix,
+                        const Matx33f &registeredCameraMatrix,
+                        const Mat_<float> &registeredDistCoeffs,
+                        const Matx44f &rbtRgb2Depth,
+                        const Size outputImagePlaneSize,
+                        const bool depthDilation,
+                        const float inputDepthToMetersScale,
+                        Mat &registeredDepth)
+    {
+
+        // Create output Mat of the correct type, filled with an initial value indicating no depth
+        registeredDepth = Mat_<DepthDepth>(outputImagePlaneSize, noDepthSentinelValue<DepthDepth>());
+
+        // Figure out whether we'll have to apply a distortion
+        bool hasDistortion = (countNonZero(registeredDistCoeffs) > 0);
+
+        // A point (i,j,1) will have to be converted to 3d first, by multiplying it by K.inv()
```

Comment:
Can be re-used  depth utilities already in the [rgbd module](https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/src/depth_to_3d.cpp)?